File: convert/Stellaris.Convert/Stellaris.Convert/Parser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Stellaris.Convert
{
    public class Parser
    {
        private Lexer lexer;
        private Token token;

        public Parser(Lexer lexer)
        {
            this.lexer = lexer;
        }

        public object Parse()
        {
            this.token = this.lexer.Token();

            var pairs = new List<Tuple<string, object>>();
            while (!this.token.Type.Equals("eof"))
            {
                pairs.Add(this.ParsePair());
            }

            return this.ConvertPairs(pairs);
        }

        private Tuple<string, object> ParsePair()
        {
            if (this.token.Type.Equals("text"))
            {
                var keyOrValue = this.token.Value;

                this.token = this.lexer.Token();
                if (this.token.Type.Equals("="))
                {

                    this.token = this.lexer.Token();
                    if (this.token.Type.Equals("text"))
                    {
                        var value = this.token.Value;
                        this.token = this.lexer.Token();
                        return new Tuple<string, object>(keyOrValue, value);
                    }
                    else if (this.token.Type.Equals("{"))
                    {
                        return new Tuple<string, object>(keyOrValue, this.ParseObject());
                    }
                    else
                    {
                        throw new InvalidOperationException();
                    }

                }
                else
                {
                    return new Tuple<string, object>(null, keyOrValue);
                }
            }
            else if (this.token.Type.Equals("{"))
            {
                return new Tuple<string, object>(null, this.ParseObject());
            }
            else
            {
                throw new InvalidOperationException();
            }
        }

        private object ParseObject()
        {
            this.token = this.lexer.Token();

            var pairs = new List<Tuple<string, object>>();
            while (!this.token.Type.Equals("}"))
            {
                pairs.Add(this.ParsePair());
            }

            this.token = this.lexer.Token();

            return this.ConvertPairs(pairs);
        }

        private object ConvertPairs(List<Tuple<string, object>> pairs)
        {
            if (!pairs.Any())
            {
                return pairs;
            }

            var isList = pairs.All(t => t.Item1 == null);
            if (isList)
            {
                return pairs
                    .Select(t => t.Item2)
                    .ToList();
            }
            else
            {
                return pairs
                    .GroupBy(t => t.Item1, t => t.Item2)
                    .ToDictionary(
                        g => g.Key,
                        g => (g.Count() == 1) ? g.Single() : g.ToList()
                    );
            }
        }
    }
}

File: stellaris/parse/parser.py
class Parser:
    def __init__(self, lexer):
        self._lexer = lexer
        self._token = None

    def parse(self):
        self._token = self._lexer.token()

        pairs = []
        while self._token[0] != "eof":
            pairs.append(self._parse_pair())
        return self._convert_pairs(pairs)

    def _parse_pair(self):
        if self._token[0] == "text":
            key_or_value = self._token[1]
            self._token = self._lexer.token()
            if self._token[0] == "=":
                self._token = self._lexer.token()
                if self._token[0] == "text":
                    value = self._token[1]
                    self._token = self._lexer.token()
                    return (key_or_value, value)
                elif self._token[0] == "{":
                    return (key_or_value, self._parse_object())
                else:
                    raise Exception()    
            else:
                return (None, key_or_value)
        elif self._token[0] == "{":
            return (None, self._parse_object())

    def _parse_object(self):
        self._token = self._lexer.token()
        pairs = []
        while self._token[0] != "}":
            pairs.append(self._parse_pair())
        self._token = self._lexer.token()
        return self._convert_pairs(pairs)

    def _convert_pairs(self, pairs):

        if len(pairs) == 0:
            return pairs

        groups = {}
        is_list = True
        for (key, value) in pairs:
            if key not in groups:
                groups[key] = [value]
            else:
                groups[key].append(value)

            if key != None:
                is_list = False

        if is_list:
            return [value for (key,value) in pairs]
        else:
            for key in groups.keys():
                if len(groups[key]) == 1:
                    groups[key] = groups[key][0]
            return groups